packages/mds-collector-backend/repository/index.ts

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright 2021 City of Los Angeles
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { InsertReturning, ReadWriteRepository, RepositoryError } from '@mds-core/mds-repository'
+import { CollectorMessageDomainCreateModel, CollectorMessageDomainModel } from '../@types'
+import { CollectorMessageEntity, CollectorMessageEntityModel } from './entities'
+import { CollectorMessageDomainToEntityCreate, CollectorMessageEntityToDomain } from './mappers'
+import migrations from './migrations'
+
+class CollectorReadWriteRepository extends ReadWriteRepository {
+  public insertCollectorMessages = async (
+    messages: CollectorMessageDomainCreateModel[],
+    beforeCommit: () => Promise<void> = async () => undefined
+  ): Promise<CollectorMessageDomainModel[]> => {
+    try {
+      const connection = await this.connect('rw')
+
+      const chunks = this.asChunksForInsert(
+        messages.map(CollectorMessageDomainToEntityCreate.mapper({ recorded: Date.now() }))
+      )
+
+      const results: Array<InsertReturning<CollectorMessageEntityModel>> = await connection.transaction(
+        async manager => {
+          const committed = await Promise.all(
+            chunks.map(chunk =>
+              manager
+                .getRepository(CollectorMessageEntity)
+                .createQueryBuilder()
+                .insert()
+                .values(messages.map(CollectorMessageDomainToEntityCreate.mapper()))
+                .returning('*')
+                .execute()
+            )
+          )
+          await beforeCommit()
```

Comment:
I think that having this closure is probably a little excessive... Is there really much value in creating the transaction, then running a closure (as used so far, to write to Kafka), then committing the transaction? I don't see much value that is added as opposed to just running the Kafka writes after properly persisting the data in PG (or beforehand, if you don't care about it being persisted in PG); and the usage looks pretty weird (especially given that JS doesn't have named parameters), it seems as if it's a CB function but it.. isn't really?

Reply:
The idea was that the PG and Kafka writes both succeed so the Kafka writes are done within the transaction. I could change the second parameter to `options: Partial<{ beforeCommit: () => Promise<void> }> = {}`